man/Repository/bkd.man
.so ../macros
.TH bkd Repository %G% "BitMover, Inc." BitKeeper
.\"	============== The BitKeeper daemon ==============
.SH NAME
bk bkd \- the BitKeeper daemon
.SH SYNOPSIS
.if t \{\
.B bk bkd [\-eid] [\-l
.IB log ]
.B [\-P
.IB file ] 
.B [\-p
.IB port ] 
.B [\-t
.IB num ] 
.B [\-u
.IB uid ] 
.B [\-x
.IB cmd ]
.\}
.if n
bk bkd [-eid] [-l<log>] [-P<file>] [-p<port>] [-t<num>] [-u<uid>] [-x<cmd>]
.SH DESCRIPTION
.LP
The BitKeeper daemon, bkd, is used to get information about a
repository from a remote machine.  It is run in one of two ways,
.sp
.RS .5i
a) via rsh/ssh, i.e., "rsh remote host bkd",
.br
b) as a stand\-alone daemon, i.e., "bkd \-d".
.RE
.LP
The method used depends on how the remote repository is named.
See bk help parent for details on the naming syntax.
.LP
The preferred method is via ssh, since ssh provides a secure channel.
.LP
The stand\-alone daemon method currently has no security, other than the
ability to run in read\-only mode.  The daemon does support compression.
.SH COMMANDS
.LP
The commands are not intended for user access; BitKeeper client programs
interact with the daemon directly.  However, it may be useful to access
the commands directly for debugging.  It is possible to telnet to the
BitKeeper port, in the same way it is possible to telnet to the SMTP port.
.LP
The command summary is below.  The options tend to track the client
commands fairly closely.
.sp
.RS .5i
cd pathname
.br
clone [\-q] [\-z[0\-9]]
.br
help
.br
pull [\-lnq] [\-z[0\-9]]
.br
push [\-q] [\-z[0\-9]]
.br
pwd
.br
rootkey
.br
status [\-v]
.br
version
.RE
.LP
The client commands parse their options, connect to the server, and
send the appropriate commands to complete the action.  When started
with the \-l option, all sent commands are logged.
.SH "ANONYMOUS ACCESS"
.LP
The most common use of the stand\-alone daemon is for anonymous access
to a repository.  To provide read\-only, anonymous access, you can run:
.AX
bk bkd \-d \-xpush
.XA
.LP
This will allow anyone to read (but not write) all repositories on the
system running the daemon.
.LP
If you want to export a single repository, pick a port number, and 
do this:
.AX
cd /u/linux
bk bkd \-d \-p5555 \-xcd \-xpush
.XA
.LP
This says to run in daemon mode, bind to port 5555, and disallow the
"cd" and "push" commands.  By disallowing the "cd" command, the daemon
at port 5555 is tied to the repository in the current working directory
(bkd needs to be run at the root of the repository).  By disallowing the
"push" command, the repository is protected from updates.
.LP
Clients can get to this repository by using the BK URL of
.AX    
bk://host.domain:5555
.XA
.br
i.e.,
.AX
$ bk clone bk://host.domain:5555 my_tree
.XA
.LP
.SH OPTIONS
.TP
.B \-d
Run as a daemon.
.TP
.B \-e
Exit on any error.
.TP
.B \-i
Interactive, provide prompts.
.TP
.AN \-l log
Log accesses in 
.IR log ;
if log is not specified, then to stderr.
.TP
.AN \-P pfile
Write the pid of daemon process into this file at startup.
.TP
.AN \-p port
Specify an alternative port.  The default port is currently
0x3962 (aka 14690) but that will change once we get a
reserved port. 
.TP
.AN \-t num
Exit after 
.I num 
seconds. (used when debugging)
.TP
.AN \-u uid
Run as UID
.IR uid .
.TP
.AN \-x cmd
Exclude 
.I cmd 
from the allowed command list.
.SH EXAMPLES
.LP
We use the following in /var/bitkeeper/repositories to provide anonymous 
read only access to some BitKeeper repositories.  This works with the
init script shown below:
.AX
#----------------------\ cut\ here\ --------------------------
/home/bk/LMbench \-p5000 \-xcd \-xpush \-u99
/home/bk/bitcluster \-p6000 \-xcd \-xpush \-u99
/home/bk/one \-p7000 \-xcd \-xpush \-u99
#----------------------\ cut\ here\ --------------------------
.XA
.LP
We use the following as /etc/rc.d/init.d/bitkeeper to start up BitKeeper
on bitkeeper.com:
.AX
#----------------------\ cut\ here\ --------------------------
#!/bin/sh
#
# bitkeeper	Start/stop the bitkeeper daemon.
# 		@(#)bitkeeper.init 1.1 Copyright (c) 2000 Larry McVoy

# Source networking configuration.
if [ \-f /etc/sysconfig/network ]
then	. /etc/sysconfig/network

	# Check that networking is up.
	[ ${NETWORKING} = "no" ] && exit 0
fi
[ \-x /usr/bin/bk ] || exit 0
VAR=/var/bitkeeper

case "$1" in
    start_msg)	echo "Start BitKeeper daemons"
		;;
    stop_msg)	echo "Stop BitKeeper daemons"
		;;
    start)	cd $VAR || exit 1
		test \-f repositories || {
			echo Nothing advertised
			exit 0
		}
		while read dir opts
		do	(
			cd $dir || exit 1
			F=`basename $dir`
			bk bkd $opts \-l$VAR/log.$F \-P$VAR/pid.$F
			echo Started bkd $opts in $dir
			)
		done < repositories
		;;

    stop)	
		cd $VAR || exit 1
		echo Shutting down BitKeeper daemons
		for i in pid.*
		do	kill \-HUP `cat $i`
			rm $i
		done
		;;

    status)	ps \-axf | grep bkd
		;;

    *)		echo "Usage: bitkeeper {start|stop}"
		exit 1
		;;
esac
exit 0
#----------------------\ cut\ here\ --------------------------
.XA
.SH BUGS
.LP
The log files are currently single user because they don't append.
.br
Uses one log file per daemon.
.br
Needs password support.
.br
Needs access control based on authenticated users.
.SH "SEE ALSO"
.B bk help qs_bkd
.br
.B bk help parent
.br
.B bk help initscripts
.SH ALIASES
.BR anonymous (Repository)
.br
.BR deamon (Repository)
.br
.BR daemon (Repository)
.br
.BR demon (Repository)
.br
.BR security (Repository)


RELEASE-NOTES
============================================================================
Release notes for BitKeeper version 6.0.2 (released Oct 29, 2013)

    This is a bugfix release (one feature snuck in).

Bugfixes

  bk diffs
    - Fixed a bug where 'bk diffs -NL' did not list extra files.
    - bk diffs -s was not working because of the switch to -u as
      the default.
    - Function header matching in 'bk diffs -p' now works with C++
      names.

  bk difftool
    - When looking at diffs over different types of files, a binary
      file in the list would cause diffs for the next text file to not
      be displayed.

  bk gone
    - People would try and run 'bk gone foo.c' which is incorrect, the
      correct usage is:
	bk gone `bk log -r+ -nd:ROOTKEY:`
      bk gone now works harder to validate that the arguments passed
      are actually keys.

  Other bugfixes
    - Fix a bug in the BK6->BK5 compat code where some files were not
      converted correctly to the BK5 format.
    - Stale locks on windows were not always being cleaned immediately.
    - Ignore extra files in .bk which were causing problems in Mac OS.

Features

   bk annotate
    - MD5 keys are now supported in the annotation options by running
      either 'bk annotate -a5' or 'bk annotate -A5'.

============================================================================
Release notes for BitKeeper version 6.0.1 (released Aug 9, 2013)

    This is a bugfix release (a few features snuck in).

Bugfixes

  bk clone
    - 'bk clone -@' wasn't populating comps when the remote repository
      was a subset of the local baseline.
    - 'bk clone -r<rev> [--upgrade|--downgrade] repo localcopy' could
      fail in some cases, and has been fixed.
    - 'bk clone -r<rev> bk://host/repo localcopy' will now fail if the
      BK client version is bk-6.0 and the bkd is bk-6.0.1 or later.
      The solution is to upgrade the client.

  bk commit
    - 'bk commit -f' was still prompting for comments.
    - 'bk commit' had lost some status reporting; that has been
      restored.

  bk collapse
    - 'bk collapse' would leave components marked as pending; this caused
      pull to error with a "KEY MISMATCH" message.
    - 'bk fix/collapse' in a gate was only partially disabled, it would
      collapse in the product but would refuse to collapse in
      components.

  bk diffs
    - 'bk -Ucx diffs -N' would only print diffs for one new file;
      fixed to print diffs for all new files

  bk difftool
    - The file listing box on difftool resizes itself to be as large
      as the longest pathname shown (it used to truncate).
    - 'bk difftool -S -Lpath/to/component' is now supported.
      (-S == scan only the current component)
    - The 'next' button in bk difftool didn't come up highlighted when
      there was more than one file.
    - Better error reporting with -L when a single GCA doesn't exist.

  bk pull
    - A pull from a remote repository served by a bk version 5.x could fail
      if a component clone was part of the operation.  It works now.

  bk push
    - A push from a local repository served by a bk version 5.x to a remote
      repository that uses new storage format of bk-6 could fail if a component
      clone was part of the operation.  It works now.

  bk revtool
    - A particular sequence of clicks and launching difftool could
      make revtool unresponsive.

    - When launching revtool with a line number (bk revtoo +lno),
      the comments for the revision that added that line were not being shown.

  bk sfiles
    - In rare cases, a pending component would not be seen as pending
      ('bk -pA' did not list them but it does now).

  Misc
    - In rare cases, permissions in the .bk/SCCS directory could be set
      wrong and cause bk to fail.  The known case has been fixed and
      the problem is now repaired automatically.

    - In another permissions failure, when a user didn't have write
      permissions on the repository, commit would fail with an
      unhelpful error message.  The problem is now reported more
      clearly.

  Graphical tools
    - When resizing graphical tools quickly, the resize labels could
      get out of sync; they update more frequently now.

Features
    - bk difftool now supports -w/-b to ignore white space changes
      like 'bk diffs'.  There is toolbar control of this feature as
      well.

    - BitKeeper versions prior to 6.0 may now clone repositories in the new
      format; they will automatically be downgraded to a compatible format.

============================================================================
Release notes for BitKeeper version 6.0 (released Jul 05, 2013)

    The primary focus of the 6.0 effort is performance improvements.

New repository format for performance
    BK6 has a new file format designed for performance.

    The new repository format cannot be read by bk-5.x releases
    of BitKeeper, but clone can convert a new repository to a bk-5.x
    readable repository.

Compatibility
    New file formats are a cause for concern but we've done the extra
    work so you may try it without forcing everyone to upgrade. BK6 can
    pull/push the new format from/to bk5's format.

    To enable the the new, faster format:
  
	bk clone --upgrade bk5-repo bk6-repo
  
    To downgrade to be compatible with bk5 (or bk4 if it is a non-nested
    repo):

	bk clone --downgrade bk6-repo bk5-repo

    To allow clones/pulls/pushes from bk4 & bk5, serve the repositories
    with a bk6 bkd:

    	cd bk6-repo
	bk pull bk://bk6-server/bk5-repo
	bk push bk://bk6-server/bk5-repo
 
Performance improvements:
    In the examples below a non-nested test repository contained 55,000
    files with 282,000 changesets on an NFS server.  Both the server
    and client caches were dropped (made cold) before each operation.

    Note that none of the performance improvements depend on the use of
    a file system notifier.  We considered and rejected such an approach

src/bkd.c
/*
 * Copyright 1999-2016 BitMover, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "bkd.h"

#define	LOG_STDERR	(char *)1

private	void	exclude(char *cmd, int verbose);
private void	unexclude(char **list, char *cmd);
private	int	findcmd(int ac, char **av);
private	int	getav(FILE *f, int *acp, char ***avp);
private	void	log_cmd(char *peer, int ac, char **av, int);
private	int	do_cmds(void);
private int	svc_uninstall(void);
private int	nextbyte(void *unused, char *buf, int size);

char		*bkd_getopt = "a:cCdDeE:hi:l|p:P:qRSt:UV:x:";
private char	**exCmds;

/*
 * An easy way of debugging bkd commands (bkd_*.c files - cmd_functions)
 * is to run them directly from the command line (perhaps under gdb).
 * E.g. for debugging pull:
 *
 * $ BK_FEATURES=SAMv3,BAMv2 BK_REMOTE_PROTOCOL=1.3 ./bk -P _bkd_pull_part1
 *
 * Then type the protocol directly.
 */
int
bkd_main(int ac, char **av)
{
	int	port = 0, daemon = 0, check = 0;
	char	*addr = 0, *p;
	int	i, c;
	char	**unenabled = 0;
	char	**args = 0;

	bzero(&Opts, sizeof(Opts));	/* just in case */

	/*
	 * Any commands off by default should be added here like so:
	 *	unenabled = addLine(unenabled, "dangerous_command");
	 * Note the freeLines below does not free the line itself.
	 */
	unenabled = addLine(unenabled, "kill");

	/*
	 * -c now means check arguments and if they are OK echo the starting
	 * dir and exit.  Used for service.
	 */
	while ((c = getopt(ac, av, bkd_getopt, 0)) != -1) {
		/*
		 * Note that bkd_level.c depends on each arg being distinct,
		 * not pushed together.  This is fine, don't change it.
		 */
		args = addLine(args, aprintf("-%c%s", c, optarg ? optarg : ""));
		switch (c) {
		    case 'a': Opts.portfile = strdup(optarg); break;
		    case 'c': check = 1; break;
		    case 'C': Opts.safe_cd = 1; break;		/* doc */
		    case 'd': daemon = 1; break;		/* doc 2.0 */
		    case 'D': Opts.foreground = 1; break;	/* doc 2.0 */
		    case 'i':                                   /* undoc 2.0 */
			if (streq(optarg, "kill")) Opts.kill_ok = 1;
			unexclude(unenabled, optarg);
			break;
		    case 'h': Opts.http_hdr_out = 1; break;	/* doc 2.0 */
		    case 'l':					/* doc 2.0 */
			Opts.logfile
			    = optarg ? fullname(optarg, 0) : LOG_STDERR;
			break;
		    case 'V':	/* XXX - should be documented */
			Opts.vhost_dirpath = strdup(optarg); break;
		    case 'p':
			if ((p = strchr(optarg, ':'))) {
				*p = 0;
				addr = strdup(optarg);
				*p = ':';
				port = atoi(++p);
			} else {
				port = atoi(optarg);
				addr = 0;
			}
			break;	/* doc 2.0 */
		    case 'P': Opts.pidfile = optarg; break;	/* doc 2.0 */
		    case 'S': 					/* undoc 2.0 */
			Opts.symlink_ok = 1;
			break;
		    case 'R': 					/* doc 2.0 */
			if (getenv("BKD_SERVICE")) break;	/* ignore it */
			return (svc_uninstall());
		    case 'x': exclude(optarg, 1); break;	/* doc 2.0 */
		    case 'e': break;				/* obsolete */
		    case 'E': putenv(optarg); break;		/* undoc */
		    case 'q': Opts.quiet = 1; break; 		/* undoc */
		    case 't': Opts.alarm = atoi(optarg); break;	/* undoc */
		    case 'U': Opts.unsafe = 1; break;
		    default: bk_badArg(c, av);
	    	}
	}
	EACH(unenabled) exclude(unenabled[i], 0);
	freeLines(unenabled, 0);
	if ((port || check) && Opts.portfile) usage();
	if (av[optind] && !getenv("BKD_SERVICE")) usage();

	if (av[optind] && chdir(av[optind])) {
		perror(av[optind]);
		exit(1);
	}

	unless (p = joinLines(" ", args)) p = strdup("");
	freeLines(args, free);
	safe_putenv("_BKD_OPTS=%s", p);
	free(p);

	unless (Opts.vhost_dirpath) Opts.vhost_dirpath = strdup(".");

	if (port || addr || Opts.portfile) daemon = 1;
	if (daemon && (Opts.logfile == LOG_STDERR) && !Opts.foreground) {
		fprintf(stderr, "bkd: Can't log to stderr in daemon mode\n");
		return (1);
	}
	core();
	putenv("PAGER=cat");
	putenv("_BK_IN_BKD=1");
	if (daemon) {
		unless (Opts.portfile) {
			unless (port) port = BK_PORT;
			if ((c = tcp_connect("127.0.0.1", port)) > 0) {
				unless (Opts.quiet) {
					fprintf(stderr, "bkd: "
					    "localhost:%d is already in use.\n",
					    port);
				}
				closesocket(c);
				return (2);	/* regressions count on 2 */
			}
			if (check) return (0);
		}
		safe_putenv("_BKD_PORT=%d", port);
		safe_putenv("_BKD_ADDR=%s", addr ? addr : "0.0.0.0");
		if (addr) free(addr);
		if (Opts.logfile && (Opts.logfile != LOG_STDERR)) {
			safe_putenv("_BKD_LOGFILE=%s", Opts.logfile);
		}
		bkd_server(ac, av);
		exit(1);
		/* NOTREACHED */
	} else {
		callstack_push(0);
		i = do_cmds();
		callstack_pop();
		T_CMD("bk bkd = %d", i);
		return (i);
	}
}

private void
unexclude(char **list, char *cmd)
{
	if (removeLine(list, cmd, 0) == 1) return;
	fprintf(stderr, "bkd: %s was not excluded.\n", cmd);
	exit(1);
}

/*
 * Shutdown the socket connection to the bk client when we decide to
 * exit early without reading all stdin.  This implements Apache's
 * "lingering close" to send an EOF to the client and avoid sending a
 * RST from our kernel if we fail to consume all stdin.
 */
void
drain(void)
{
	char	buf[1024];
	int	i = 0;

	signal(SIGALRM, exit);
	alarm(20);
	shutdown(1, 1);		/* send an EOF to the client */
	close(1); /* in case remote is waiting for input */
	/* consume all stdin */
	while (read(0, buf, sizeof(buf)) > 0) {
		if (i++ > 200) break; /* just in case */
	}
	shutdown(0, 2);		/* now fully shutdown */
	closesocket(0);
}

off_t
get_byte_count(void)
{

	char buf[MAXPATH];
	off_t	byte_count = 0;
	FILE *f = 0;

	unless (proj_root(0)) return (0);
	concat_path(buf, proj_root(0), "/BitKeeper/log/byte_count");
	f = fopen(buf, "r");
	if (f && fgets(buf, sizeof(buf), f)) {
		if (strlen(buf) > 11) {
			fprintf(stderr, "Holy big transfer, Batman!\n");
			fclose(f);
			return ((off_t)0xffffffff);
		}
		byte_count = strtoul(buf, 0, 10);
	}
	if (f) fclose(f);
	return (byte_count);
}


void
save_byte_count(unsigned int byte_count)
{
	FILE	*f;
	char	buf[MAXPATH];

	unless (proj_root(0)) return;
	concat_path(buf, proj_root(0), "/BitKeeper/log/byte_count");
	f = fopen(buf, "w");
	if (f) {
		fprintf(f, "%u\n", byte_count);
		fclose(f);
	}
}

private int
do_cmds(void)
{
	int	ac;
	char	**av;
	int	i, ret = 1, log;
	int	debug = getenv("BK_DEBUG") != 0;
	char	*peer = 0;
	int	logged_peer = 0;
	FILE	*f = 0;

	f = funopen(0, nextbyte, 0, 0, 0);
	unless (peer = getenv("BKD_PEER")) {
		peer = "local";
		logged_peer = 1;
	}

	if (Opts.alarm) {
		signal(SIGALRM, exit);
		alarm(Opts.alarm);
	}
	/* Don't allow existing env to be used */
	/* XXX put function in utils.c and have it match what gets set? */
	putenv("BK_REMAP=");
	putenv("BK_FEATURES=");
	putenv("BK_FEATURES_REQUIRED=");
	putenv("BK_QUIET_TRIGGERS=");
	putenv("BK_SYNCROOT=");


	/*
	 * The _BKD_HTTP varible indicates that the current bkd
	 * connection was initiated by a http POST command and so the
	 * client will be disconnecting after each response from the
	 * bkd.  This is used to control some aspects of locking, but
	 * should be used sparingly as it makes the logic harder to
	 * follow than if the client just specified what to do in the
	 * command stream.
	 *
	 * This varible is valid inside cmd_*() functions and in
	 * subprocesses spawned from those functions.
	 */
	putenv("_BKD_HTTP=");

	/*
	 * We only want locks from the protocol, not from process
	 * inheritance
	 */
	putenv("_BK_NESTED_LOCK=");
	putenv("BKD_NESTED_LOCK=");

	while (getav(f, &ac, &av)) {
		if (debug) {
			for (i = 0; av[i]; ++i) {
				ttyprintf("[%2d] = %s\n", i, av[i]);
			}
		}
		getoptReset();
		if ((i = findcmd(ac, av)) != -1) {
			if (Opts.logfile) log_cmd(peer, ac, av, 0);
			if (Opts.http_hdr_out) http_hdr();
			log = !streq(av[0], "putenv") && !streq(av[0], "cd");
			if (log) cmdlog_start(av, 1);

			/*
			 * Do the real work
			 */
			ret = cmds[i].cmd(ac, av);
			fflush(stdout);
			if (peer && !logged_peer) {
				/* first command records peername */
				if (log) cmdlog_addnote("peer", peer);
				logged_peer = 1;
			}
			if (debug) ttyprintf("cmds[%d] = %d\n", i, ret);

			if (log && cmdlog_end(ret, 1)) break;
			if (ret != 0) {
				out("ERROR-exiting\n");
				break;
			}
		} else if (av[0]) {
			EACH(exCmds) {
				if (streq(av[0], exCmds[i])) break;
			}
			if (Opts.logfile) log_cmd(peer, ac, av, 1);
			out("ERROR-BAD CMD: ");
			out(av[0]);
			if (i < nLines(exCmds)) {
				out(" is disabled for this bkd using -x\n");
			} else {
				out(", Try help\n");
			}
		} else {
			out("ERROR-Try help\n");
		}
	}
	getav(0, 0, 0);	/* free internal mem */
	fclose(f);
	repository_unlock(0, 0);
	drain();
	return (ret);
}

private	void
log_cmd(char *peer, int ac, char **av, int badcmd)
{
	time_t	t;
	struct	tm *tp;
	int	i;
	FILE	*log;
	static	char	**putenv;

	if (streq(av[0], "putenv")) {
		putenv = addLine(putenv, strdup(av[1]));
		return;
	}

	log = (Opts.logfile == LOG_STDERR) ? stderr : fopen(Opts.logfile, "a");
	unless (log) return;

	time(&t);
	tp = localtimez(&t, 0);
	if (putenv) {
		fprintf(log, "%s %.24s ", peer, asctime(tp));
		sortLines(putenv, 0);
		EACH(putenv) {
			if (strneq("_BK_", putenv[i], 4) ||
			    strneq("BK_", putenv[i], 3)) {
			    	fprintf(log, "%s ", &putenv[i][3]);
			} else {
			    	fprintf(log, "%s ", putenv[i]);
			}
		}
		fprintf(log, "\n");
		freeLines(putenv, free);
		putenv = 0;
	}
	fprintf(log, "%s %.24s ", peer, asctime(tp));
	if (badcmd) fprintf(log, "BAD CMD: ");
	for (i = 0; i < ac; ++i) {
		fprintf(log, "%s ", av[i]);
	}
	fprintf(log, "\n");
	unless (log == stderr) {
		fclose(log);
		log_rotate(Opts.logfile);
	}
}

/*
 * Remove any command with the specfied prefix from the command array
 */
private	void
exclude(char *cmd_prefix, int verbose)
{
	struct	cmd c[100];
	int	i, j, len;
	int	foundit = 0;

	for (i = 0; cmds[i].name; i++);
	assert(i < 99);
	for (i = j = 0; cmds[i].name; i++) {
		len = strlen(cmd_prefix);
		unless ((strlen(cmds[i].realname) >= len) &&
			strneq(cmd_prefix, cmds[i].realname, len)) {
			c[j++] = cmds[i];
		} else {
			exCmds = addLine(exCmds, strdup(cmds[i].name));
			foundit++;
		}
	}
	unless (foundit) {
		unless (verbose) return;
		fprintf(stderr, "bkd: command '%s' not found.\n", cmd_prefix);
		exit(1);
	}
	for (i = 0; i < j; i++) {
		cmds[i] = c[i];
	}
	cmds[i].name = 0;
	cmds[i].realname = 0;
	cmds[i].cmd = 0;
}

private	int
findcmd(int ac, char **av)
{
	int	i;
	char	*av0;

	if (ac == 0) return (-1);
	av0 = av[0];
	for (i = 0; cmds[i].name; ++i) {
		if (strcasecmp(av[0], cmds[i].name) == 0) {
			if (streq(av[0], "pull")) av[0] = "remote pull";
			if (streq(av[0], "push")) av[0] = "remote push";
			if (streq(av[0], "clone")) av[0] = "remote clone";
			if (streq(av[0], "abort")) av[0] = "remote abort";
			if (streq(av[0], "pull_part1")) {
				av[0] = "remote pull part1";
			}
			if (streq(av[0], "pull_part2")) {
				av[0] = "remote pull part2";
			}
			if (streq(av[0], "push_part1")) {
				av[0] = "remote push part1";
			}
			if (streq(av[0], "push_part2")) {
				av[0] = "remote push part2";
			}
			if (streq(av[0], "push_part3")) {
				av[0] = "remote push part3";
			}
			if (streq(av[0], "rclone_part1")) {
				av[0] = "remote rclone part1";
			}
			if (streq(av[0], "rclone_part2")) {
				av[0] = "remote rclone part2";
			}
			if (streq(av[0], "rclone_part3")) {
				av[0] = "remote rclone part3";
			}
			if (streq(av[0], "chg_part1")) {
				av[0] = "remote changes part1";
			}
			if (streq(av[0], "chg_part2")) {
				av[0] = "remote changes part2";
			}
			if (streq(av[0], "quit")) av[0]= "remote quit";
			if (streq(av[0], "rdlock")) av[0]= "remote rdlock";
			if (streq(av[0], "rdunlock")) av[0]= "remote rdunlock";
			if (streq(av[0], "nested")) av[0] = "remote nested";
			if (streq(av[0], "wrlock")) av[0]= "remote wrlock";
			if (streq(av[0], "wrunlock")) av[0]= "remote wrunlock";
			if (av0 != av[0]) {
				free(av0);
				av[0] = strdup(av[0]);
			}
			prog = av[0];
			return (i);
		}
	}
	return (-1);
}


static	int	content_len = -1;	/* content-length from http header */

private int
nextbyte(void *unused, char *buf, int size)
{
	int	ret;

	if (content_len == 0) return (0);
	--content_len;
	if ((ret = bkd_getc()) == EOF) return (0);
	buf[0] = (char)ret;
	buf[1] = 0;
	return (1);	/* returning 1 char */
}

private	int
getav(FILE *f, int *acp, char ***avp)
{
#define	QUOTE(c)	(((c) == '\'') || ((c) == '"'))
	static	char	**av;
	char	*buf, *s;
	int	i, inspace, inQuote;
	int	ac;
	FILE	*ftmp;

nextline:
	if (av) {
		freeLines(av, free);
		av = 0;
	}
	unless (f && (buf = fgetline(f)) && *buf) return (0);

	/*
	 * XXX TODO need to handle escaped quote character in args
	 *     This can be done easily with shellSplit()
	 *     Unfortunately older bk's don't always quote everything.
	 *     Current bk's should work with shellSplit().
	 */
	inspace = 1;
	inQuote = 0;
	for (i = 0; buf[i]; i++) {
		if (inQuote) {
			if (QUOTE(buf[i])) {
				av = addLine(av, fmem_close(ftmp, 0));
				ftmp = 0;
				inQuote = 0;
				s = 0;
			} else {
				/* a backslash inside a quoted string
				 * escapes the next character.
				 */
				if (buf[i] == '\\') ++i;
				fputc(buf[i], ftmp);
			}
			continue;
		}
		if (QUOTE(buf[i])) {
			assert(!QUOTE(buf[i+1])); /* no null args */
			ftmp = fmem();
			inQuote = 1;
			continue;
		}
		if (isspace(buf[i])) {
			buf[i] = 0;
			if (s) {
				av = addLine(av, strdup(s));
				s = 0;
			}
			inspace = 1;
		} else if (inspace) {
			s = buf + i;
			inspace = 0;
		}
	}
	if (s) {
		av = addLine(av, strdup(s));
		s = 0;
	}

	/* end of line */
	av = addLine(av, 0);
	ac = nLines(av);

	/*
	 * Process http post command used in http based
	 * push/pull/clone Strip the http header so we can access the
	 * real push/pull/clone command
	 */
	if ((ac >= 1) && streq("POST", av[1])) {
		remote	r;

		bzero(&r, sizeof (remote));
		r.wfd = 1;
		/* if stdin used, then have read_blk() use stdin */
		if (Opts.use_stdio) r.rf = f;
		skip_http_hdr(&r);
		content_len = r.contentlen;
		http_hdr();
		putenv("_BKD_HTTP=1");
		goto nextline;
	}

	*acp = ac;
	*avp = av + 1;
	return (1);
}

private int
svc_uninstall(void)
{
	int	status = sys("bk", "service", "uninstall", "-a", SYS);

	assert(WIFEXITED(status));
	return (WEXITSTATUS(status));
}
